### PR TITLE
EVG-6850 add --commits flag to commit queue

### PR DESCRIPTION
**Pull request by ablack12**
This has been tested in staging, both with a single commit and a range.

Takes a range of <commit_hash>..<commit_hash> or an actual hash. Not sure if we need to enforce this beyond it failing if it's given input that doesn't work? I tried to detail it in the comments.

Does it make sense to have only one of rev and ref? I'm confused if gitMergeBase has to change with rev, or if ref might still want to be specified for this command.

### CODE REVIEW

**Comment by ablack12** [comment]
I could just leave it as --revision and make people type it out.

File: operations/patch_util.go [hunk 1]
```diff
@@ -374,26 +388,26 @@ func loadGitData(branch string, ref string, format bool, extraArgs ...string) (*
 	if len(extraArgs) > 0 {
 		statArgs = append(statArgs, extraArgs...)
 	}
-	stat, err := gitDiff(mergeBase, ref, statArgs...)
+	stat, err := gitDiff(mergeBase, ref, rev, statArgs...)
```
Comment (ybrill) [inline operations/patch_util.go]: According to the [git docs](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-emgitdiffemltoptionsgtltcommitgtltcommitgt--ltpathgt82308203) `..` and `...` have a different meaning for `git diff` and [git rev-list](https://git-scm.com/docs/git-rev-list#_description) then they do for `git log` and `git format-patch`.

File: operations/patch_util.go [hunk 2]
```diff
@@ -354,10 +354,24 @@ func getPatchDisplay(p *patch.Patch, summarize bool, uiHost string) (string, err
...
+}
+
+func formatRev(rev string) string {
+	if rev == "" {
+		return rev
+	}
+	if isCommitRange(rev) {
+		return rev
+	}
+	return fmt.Sprintf("%s^!", rev)
```
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/patch_util.go]: I saw some examples of this on StackOverflow and it does what I want, although there may be a more standard way of doing it:
https://git-scm.com/docs/git-rev-parse#Documentation/git-rev-parse.txt-emltrevgtemegemHEADem

File: operations/flags.go [hunk 3]
```diff
@@ -223,6 +223,13 @@ func addRefFlag(flags ...cli.Flag) []cli.Flag {
 	})
 }
 
+func addRevFlag(flags ...cli.Flag) []cli.Flag {
+	return append(flags, cli.StringFlag{
+		Name:  joinFlagNames(revisionFlagName, "rev"),
+		Usage: "specify commit hash (can also be a range `start..end`)",
```
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/flags.go]: We can but we would need to be explicit, so we can look for these in `formatRev`.

File: operations/patch_util.go [hunk 4]
```diff
@@ -413,29 +429,41 @@ func gitMergeBase(branch1, branch2 string) (string, error) {
...
-func gitFormatPatch(base string, ref string) (string, error) {
+func gitFormatPatch(base string, ref, rev string) (string, error) {
 	revisionRange := fmt.Sprintf("%s..%s", base, ref)
+	if rev != "" {
+		revisionRange = formatRev(rev)
+	}
 	return gitCmd("format-patch", "--keep-subject", "--no-signature", "--stdout", "--no-ext-diff", "--summary", "--binary", revisionRange)
 }
 
 // getLog runs "git log <base>...<ref>
```
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/patch_util.go]: Sure, I can modify the comment

File: operations/patch_util.go [hunk 5]
```diff
@@ -405,6 +419,8 @@ func loadGitData(branch string, ref string, format bool, extraArgs ...string) (*
 // resulting githash as string
 func gitMergeBase(branch1, branch2 string) (string, error) {
 	cmd := exec.Command("git", "merge-base", branch1, branch2)
+	// when would the output of this ever not be the same as with HEAD? (referring to the input for branch2)
```
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/patch_util.go]: In that case would we want to allow both ref and revision to be passed in?

File: operations/patch_util.go [hunk 6]
```diff
@@ -413,29 +429,41 @@ func gitMergeBase(branch1, branch2 string) (string, error) {
 }
 
 // gitDiff runs "git diff <base> <diffargs ...>" and returns the output of the command as a string
-func gitDiff(base string, ref string, diffArgs ...string) (string, error) {
+func gitDiff(base string, ref, rev string, diffArgs ...string) (string, error) {
```
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/patch_util.go]: I thought that at one point but then it seemed like how ref was used was different for a lot of them, and so then re-determining in each of these functions if we were using ref or revision seemed like overkill.

File: operations/patch_util.go [hunk 7]
```diff
@@ -354,10 +354,24 @@ func getPatchDisplay(p *patch.Patch, summarize bool, uiHost string) (string, err
...
+		return rev
+	}
+	return fmt.Sprintf("%s^!", rev)
+}
+
 // loadGitData inspects the current git working directory and returns a patch and its summary.
 // The branch argument is used to determine where to generate the merge base from, and any extra
 // arguments supplied are passed directly in as additional args to git diff.
-func loadGitData(branch string, ref string, format bool, extraArgs ...string) (*localDiff, error) {
+func loadGitData(branch, ref, rev string, format bool, extraArgs ...string) (*localDiff, error) {
```
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/patch_util.go]: If a revision range is passed in, that's the range of what we want to include in the patch, but I thought the base commit should stay the same.

[the remark this answers is not in the document]
Comment (ablack12) [reply operations/patch_util.go]: This kind of references my question from the comment: I'm not sure how ref and revision should interact together.

File: operations/patch_util.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/patch_util.go]: This kind of references my question in the comment, I'm not sure if we should be supporting ref and revision at the same time (potentially yes).

File: operations/patch_util.go (hunk 1 above)
Comment (ablack12) [reply operations/patch_util.go]: This does seem to be true for branches and not for commit hashes.

**Comment by ablack12** [comment]
@ybrill this is good to look at again, but needs to be re-tested in staging.

File: operations/patch_util.go [hunk 8]
```diff
@@ -368,100 +368,110 @@ func formatRev(rev string) string {
...
 	return &localDiff{fullPatch, stat, log, mergeBase}, nil
 }
 
-// gitMergeBase runs "git merge-base <branch1> <branch2>" and returns the
-// resulting githash as string
-func gitMergeBase(branch1, branch2 string) (string, error) {
+// gitMergeBase runs "git merge-base <branch1> <branch2>" (where branch2 can optionally be a githash)
+// and returns the resulting githash as string
+func gitMergeBase(branch1, ref, commits string) (string, error) {
+	branch2 := getFeatureBranch(ref, commits)
```
Comment (ybrill) [inline operations/patch_util.go]: My thinking is:
For the commit queue we probably want the base to be the tip of upstream master.
For regular patches: for `...` we probably want the base to be the first commit, for `..` we probably want the merge base of the two commits, and for a single commit we probably want `commit~`

File: operations/patch_util.go [hunk 9]
```diff
@@ -368,100 +368,110 @@ func formatRev(rev string) string {
...
 // loadGitData inspects the current git working directory and returns a patch and its summary.
 // The branch argument is used to determine where to generate the merge base from, and any extra
 // arguments supplied are passed directly in as additional args to git diff.
-func loadGitData(branch, ref, rev string, format bool, extraArgs ...string) (*localDiff, error) {
+func loadGitData(branch, ref, commits string, format bool, extraArgs ...string) (*localDiff, error) {
 	// branch@{upstream} refers to the branch that the branch specified by branchname is set to
 	// build on top of. This allows automatically detecting a branch based on the correct remote,
-	// if the user's repo is a fork, for example.
+	// if the user's repo is a fork, for example. This also works with a commit hash, if given.
+	// In the case a range is passed, we only need one commit to determine the base, so we use the last commit.
```
Comment (ybrill) [inline operations/patch_util.go]: it looks like we're using the first commit

File: operations/patch_util.go (hunk 1 above)
Comment (ybrill) [reply operations/patch_util.go]: In my experiments it looked like it's an issue for hashes as well.
Let's talk this over in person.

File: operations/flags.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ybrill) [reply operations/flags.go]: > For .. and ..., if and are both commits then these are the same

It didn't look this way to me when I experimented with it.

File: operations/commit_queue.go [hunk 10]
```diff
@@ -20,7 +20,8 @@ const (
 	itemFlagName        = "item"
 	pauseFlagName       = "pause"
 	resumeFlagName      = "resume"
-	revisionFlagName    = "revision"
+	commitFlagName      = "commits"
```
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/commit_queue.go]: yes sorry

File: operations/patch_util.go (hunk 9 above)
Comment (ablack12) [reply operations/patch_util.go]: Yes sorry I changed this later.

File: operations/patch_util.go (hunk 8 above)
Comment (ablack12) [reply operations/patch_util.go]: You'll have to explain what you mean to me later. Are you saying that for a single commit As far as I can tell, using commit1 for any of `<commit1>..<commit2>` and `<commit1>...<commit2>` and `<commit1>` in gitFeatureBranch seems to work fine.

Comment (ablack12) [reply operations/patch_util.go]: I forgot that regular patches touched this. But regular patches don't have a commit flag? So we don't need to handle that case, it'll just return ref like it did before.

File: operations/flags.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ablack12) [reply operations/flags.go]: (updated comment, some formatting issues before)

File: operations/commit_queue.go [hunk 11]
```diff
@@ -20,6 +20,8 @@ const (
 	itemFlagName        = "item"
 	pauseFlagName       = "pause"
 	resumeFlagName      = "resume"
+	commitFlagName      = "commit"
+	commitsFlagName     = "commits"
```
Comment (tychoish) [inline operations/commit_queue.go]: I feel like the commit/commits api is confusing and we could just have a "commit" flag specified more than once...

Comment (ablack12) [reply operations/commit_queue.go]: Switched to be just "--commits" or "-c"

File: operations/patch_util.go [hunk 12]
```diff
@@ -354,56 +354,102 @@ func getPatchDisplay(p *patch.Patch, summarize bool, uiHost string) (string, err
...
+		return nil
+	}
+
+	commitsList := strings.Split(commits, "..")
+	if len(commitsList) != 2 { // extra check
+		return errToReturn
+	}
+
+	for i := range commitsList {
+		commitsList[i] = strings.Trim(commitsList[i], ".") // handle extra dot notation
```
Comment (ybrill) [inline operations/patch_util.go]: won't the extra "." just be on the second commit?

File: operations/commit_queue.go [hunk 13]
```diff
@@ -20,6 +20,7 @@ const (
 	itemFlagName        = "item"
 	pauseFlagName       = "pause"
 	resumeFlagName      = "resume"
+	commitsFlagName     = "commits"
```
Comment (ybrill) [inline operations/commit_queue.go]: Should the new flag be added for regular patches as well?

File: operations/patch_util.go (hunk 12 above)
Comment (ablack12) [reply operations/patch_util.go]: Yeah theoretically, unless like they have four "..." or something but I figured it doesn't hurt to be safe.

File: operations/commit_queue.go (hunk 13 above)
Comment (ablack12) [reply operations/commit_queue.go]: No, I think handling that is out of scope.

**Pull request merged**